### PR TITLE
CNTRLPLANE-1622: Enable nodepool autoscaling min zero

### PR DESCRIPTION
**Pull request by jhjaggars**
## What this PR does / why we need it:

This PR enables setting Min = 0 for NodePools to allow autoscaling to and from zero replicas.

### Key Changes:

1. **API Changes**:
   - Changed `NodePoolAutoScaling.Min` from `int32` to `*int32` (required by kubeapilinter for fields with meaningful zero values)
   - Updated validation to allow `minimum: 0` (previously `minimum: 1`)
   - Added CEL validation restricting `min=0` to AWS platform only

2. **New MachineDeployment Annotation Controller**:
   - Added new controller in `hypershift-operator/controllers/machinedeployment/`
   - Queries AWS EC2 API for instance type metadata (CPU, memory, GPU)
   - Annotates MachineDeployments with capacity information required by cluster-autoscaler
   - Enabled via `--scale-from-zero-aws-creds` flag (requires AWS credentials)
   - Critical for scale-from-zero: cluster-autoscaler needs capacity annotations to make scaling decisions
   - **Smart skip logic**: Detects if CPO natively supports scale-from-zero via image labels and automatically skips reconciliation to let CPO handle it (enables seamless migration when CPO gains native support)

3. **Controller Logic Updates**:
   - Implemented `effectiveMin` logic: enforces `min=1` for non-AWS platforms even if API validation is bypassed
   - Updated autoscaler enabled condition to check platform and version support

4. **RBAC Enhancement**:
   - Added `watch` verb to cluster-autoscaler role for `infrastructure.cluster.x-k8s.io` resources
   - Allows autoscaler's informer to use watch streams instead of polling for better performance

5. **Testing**:
   - Added E2E test `TestNodePoolAutoscalingScaleFromZero` validating scale from 0→1+ nodes
   - Added unit tests for platform-specific autoscaling logic
   - Added CreateAPIUX tests validating platform restrictions
   - Test skips when `--scale-from-zero-aws-creds` not enabled (requires MachineDeployment controller)

6. **Bug Fixes from Review**:
   - Fixed test workload to target scale-from-zero NodePool with nodeSelector
   - Fixed `mutateNodePoolAutoscalingValidation` to write spec back to schema map
   - Added nil-safe pointer dereferences using `ptr.Deref()`
   - Relaxed test assertion from exactly 2 replicas to >= 1 (pods may fit on one node)

## Which issue(s) this PR fixes:

Fixes https://issues.redhat.com/browse/CNTRLPLANE-1622

## Special notes for your reviewer:

- **Scale-from-zero is AWS-only** (restricted via CEL validation and effectiveMin enforcement)
- **Requires `--enable-scale-from-zero` flag** when installing HyperShift operator
- **Requires AWS credentials** for the MachineDeployment controller to query EC2 instance types
- The new watch permission for cluster-autoscaler is optional (falls back to polling if unavailable)
- E2E test will skip in CI unless `-e2e.enable-scale-from-zero=true` is passed
- **CPO native support migration**: The MachineDeployment controller checks CPO image labels (`io.openshift.hypershift.control-plane-operator-supports-aws-autoscaling-from-zero=true`) and skips reconciliation when CPO supports it natively. The label is currently commented out in Dockerfiles until CPO implements native support - just uncomment when ready for automatic cutover.

## Checklist:
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

**Comment by bryan-cox** [comment]
@jhjaggars - Please move this PR to draft per the contributor guidelines on the repo - https://github.com/openshift/hypershift?tab=contributing-ov-file#creating-a-pull-request

File: control-plane-operator/controllers/hostedcontrolplane/v2/assets/cluster-autoscaler/role.yaml [hunk 1]
```diff
@@ -23,6 +23,7 @@ rules:
   verbs:
   - get
   - list
+  - watch
```
Comment (jhjaggars) [inline control-plane-operator/controllers/hostedcontrolplane/v2/assets/cluster-autoscaler/role.yaml:26]: This is for this informer:
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go#L951

File: api/hypershift/v1beta1/nodepool_types.go [hunk 2]
```diff
@@ -431,9 +431,11 @@ type NodePoolManagement struct {
 // NodePoolAutoScaling specifies auto-scaling behavior for a NodePool.
 // +kubebuilder:validation:XValidation:rule="self.max >= self.min", message="max must be equal or greater than min"
 type NodePoolAutoScaling struct {
-	// min is the minimum number of nodes to maintain in the pool. Must be >= 1 and <= .Max.
+	// min is the minimum number of nodes to maintain in the pool.
+	// Can be set to 0 for scale-from-zero.
+	// Must be >= 0 and <= .Max.
 	//
-	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Minimum=0
```
Comment (LiangquanLi930) [inline api/hypershift/v1beta1/nodepool_types.go:454]: So far, only AWS supports scale-up from zero, so I think it would be best to add a kubebuilder validation, similar to what’s done here:
https://github.com/openshift/hypershift/blob/dc604559b500f4ec35d1b0c43080170784323180/api/hypershift/v1beta1/nodepool_types.go#L87

File: docs/content/how-to/agent/create-agent-cluster.md [hunk 3]
```diff
@@ -661,14 +661,30 @@ clusteroperator.config.openshift.io/storage                                    4
...
+~~~
+
+> **INFO:** With `min: 0`, the cluster will scale down to zero worker nodes when no workloads require scheduling, and automatically scale up when new workloads are created.
+
+**Important considerations for scale-from-zero:**
+
+- **Cold start delay**: When scaling from 0 to 1, there will be a delay as new agents are provisioned and join the cluster
+- **Template validation**: Ensure your NodePool configuration is valid since template issues will only be discovered during the first scale-up
+- **Monitoring**: Clusters with 0 nodes are normal in scale-from-zero scenarios, not indicative of a problem
+
```
Comment (LiangquanLi930) [inline docs/content/how-to/agent/create-agent-cluster.md]: This part covers the installation for the agent platform. IMHO, we can move it to docs/content/how-to/automated-machine-management/autoscaling-from-zero.md.

File: test/e2e/autoscaling_test.go [hunk 4]
```diff
@@ -67,6 +67,8 @@ func TestAutoscaling(t *testing.T) {
 		t.Run("TestAutoscaling", testAutoscaling(ctx, mgtClient, hostedCluster, clusterOpts.NodePoolReplicas, clusterOpts.NodePoolReplicas+2))
 
 		t.Run("TestAutoscalingBalancing", testAutoscalingBalancing(ctx, mgtClient, hostedCluster, clusterOpts.NodePoolReplicas*2, additionalNP))
+
+		t.Run("TestScaleFromZero", testScaleFromZero(ctx, mgtClient, hostedCluster))
```
Comment (LiangquanLi930) [inline test/e2e/autoscaling_test.go]: If the current solution is to use [capa-annotator](https://github.com/jhjaggars/capa-annotator), then I think we can skip adding this test for now. Once CAPA has been updated, we can bring this part of the code back.

File: api/hypershift/v1beta1/nodepool_types.go (hunk 2 above)
Comment (jhjaggars) [reply api/hypershift/v1beta1/nodepool_types.go:454]: AWS isn't the only platform that supports scale up from zero.  Any provider can (and many do) support it.

File: docs/content/how-to/agent/create-agent-cluster.md (hunk 3 above)
Comment (jhjaggars) [reply docs/content/how-to/agent/create-agent-cluster.md]: On further thought, I'm not sure how helpful this documentation is at all.  How do you feel about it being omitted from this PR?

File: test/e2e/autoscaling_test.go (hunk 4 above)
Comment (jhjaggars) [reply test/e2e/autoscaling_test.go]: Yeah, I'm not sure this sort of test makes any sense with the type of change in this PR.  I'll remove it entirely.

File: api/hypershift/v1beta1/nodepool_types.go (hunk 2 above)
Comment (LiangquanLi930) [reply api/hypershift/v1beta1/nodepool_types.go:454]: Yes, IMHO, Agent, KubeVirt, Azure, and other platforms haven’t been systematically tested yet, so I think we can stick to AWS for now and add support for the other platforms once they’ve been verified.

File: docs/content/how-to/agent/create-agent-cluster.md (hunk 3 above)
Comment (LiangquanLi930) [reply docs/content/how-to/agent/create-agent-cluster.md]: look good to me

File: hypershift-operator/controllers/nodepool/capi.go [hunk 5]
```diff
@@ -707,17 +707,27 @@ func setMachineDeploymentReplicas(nodePool *hyperv1.NodePool, machineDeployment
 		// The MachineDeployment replicas field should default to a value inside the (min size, max size) range based on the autoscaler annotations
 		// so the autoscaler can take control of the replicas field.
 		//
-		// 1. if it’s a new MachineDeployment, or the replicas field of the old MachineDeployment is < min size, use min size
+		// 1. if it's a new MachineDeployment, or the replicas field of the old MachineDeployment is <= min size, use min size
 		// 2. if the replicas field of the old MachineDeployment is > max size, use max size
+		//
```
Comment (jhjaggars) [inline hypershift-operator/controllers/nodepool/capi.go:712]: I want to call this section out.  I became concerned that supporting a minimum of 0 and 1 depending upon the platform might introduce subtle bugs in the case that Min gets set to 0 somehow.  This guard will ensure that Min will be treated as no less than 1 unless we are working against AWS.  This can be removed once we feel that the platform has no bearing on what the minimum should be.

File: api/hypershift/v1beta1/nodepool_types.go [hunk 6]
```diff
@@ -86,6 +86,7 @@ type NodePool struct {
 // +kubebuilder:validation:XValidation:rule="!has(oldSelf.arch) || has(self.arch)", message="Arch is required once set"
 // +kubebuilder:validation:XValidation:rule="self.arch != 'arm64' || has(self.platform.aws) || has(self.platform.azure) || has(self.platform.agent) || self.platform.type == 'None'", message="Setting Arch to arm64 is only supported for AWS, Azure, Agent and None"
 // +kubebuilder:validation:XValidation:rule="!has(self.replicas) || !has(self.autoScaling)", message="Both replicas or autoScaling should not be set"
+// +kubebuilder:validation:XValidation:rule="!has(self.autoScaling) || self.autoScaling.min > 0 || has(self.platform.aws)", message="Scale-from-zero (autoScaling.min=0) is currently only supported for AWS platform"
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/nodepool_types.go]: Could you add a test case for CreateAPIUX?

**Review by bryan-cox — COMMENTED** [review]
Could you also add an end to end test that tests autoscaling to zero?

File: control-plane-operator/controllers/hostedcontrolplane/testdata/cluster-autoscaler/TechPreviewNoUpgrade/zz_fixture_TestControlPlaneComponents_cluster_autoscaler_role.yaml [hunk 7]
```diff
@@ -32,6 +32,7 @@ rules:
   verbs:
   - get
   - list
+  - watch
```
[the remark this answers is not in the document]
Comment (jhjaggars) [reply control-plane-operator/controllers/hostedcontrolplane/testdata/cluster-autoscaler/TechPreviewNoUpgrade/zz_fixture_TestControlPlaneComponents_cluster_autoscaler_role.yaml:34]: I think this is a false positive:
https://github.com/openshift/hypershift/pull/6975/files#diff-bb15035f378243f3b6ac2cd1419afc50596acb9512764921e54790a3cbe01601R26

File: cmd/install/install.go [hunk 8]
```diff
@@ -223,6 +238,11 @@ func (o *Options) ApplyDefaults() {
 	default:
 		o.HyperShiftOperatorReplicas = 1
 	}
+
+	// Auto-enable scale-from-zero for AWS platform if not explicitly set
+	if !o.EnableScaleFromZero {
+		o.EnableScaleFromZero = isAWSPlatformEnabled(o.PlatformsToInstall)
```
Comment (csrwng) [inline cmd/install/install.go]: I would only enable it if creds were specified as well

File: cmd/install/install.go [hunk 9]
```diff
@@ -701,26 +725,26 @@ func setupCRDs(opts Options, operatorNamespace *corev1.Namespace, operatorServic
...
-						Strategy: apiextensionsv1.WebhookConverter,
-						Webhook: &apiextensionsv1.WebhookConversion{
-							ClientConfig: &apiextensionsv1.WebhookClientConfig{
-								Service: &apiextensionsv1.ServiceReference{
-									Namespace: operatorNamespace.Name,
-									Name:      operatorService.Name,
-									Port:      ptr.To[int32](443),
-									Path:      ptr.To("/convert"),
+					// Handle conversion webhook setup
+					if opts.EnableConversionWebhook {
```
Comment (csrwng) [inline cmd/install/install.go]: unnecessary change

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+
+	return ctrl.Result{}, nil
+}
+
+// getHostedClusterForMachineDeployment retrieves the HostedCluster associated with this MachineDeployment.
+func (r *Reconciler) getHostedClusterForMachineDeployment(ctx context.Context, md *clusterv1.MachineDeployment) (*hyperv1.HostedCluster, error) {
+	// Get CAPI Cluster in the same namespace via label
+	clusterName := md.Labels["cluster.x-k8s.io/cluster-name"]
+	if clusterName == "" {
+		return nil, fmt.Errorf("MachineDeployment missing cluster.x-k8s.io/cluster-name label")
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/controller.go]: This is not something that should be retried. Whenever it gets the label you should get another event for it.

Comment (csrwng) [reply hypershift-operator/controllers/machinedeployment/controller.go]: I would just log it and return nil.

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+		Name:      clusterName,
+	}, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get CAPI Cluster %s/%s: %w", md.Namespace, clusterName, err)
+	}
+
+	// Extract HostedCluster reference from annotation
+	hcKey := cluster.Annotations[supportutil.HostedClusterAnnotation]
+	if hcKey == "" {
+		return nil, fmt.Errorf("cluster %s missing %s annotation", clusterName, supportutil.HostedClusterAnnotation)
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/controller.go]: same here, if the annotation is missing, I'd just return nil

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+	}
+
+	return result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, machineDeployment *clusterv1.MachineDeployment) (ctrl.Result, error) {
+	klog.V(3).Infof("%v: Reconciling MachineDeployment", machineDeployment.Name)
+
+	// Check if CPO natively supports scale-from-zero
+	hc, err := r.getHostedClusterForMachineDeployment(ctx, machineDeployment)
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/controller.go]: I would handle a nil err and nil cluster by just returning.

File: hypershift-operator/controllers/machinedeployment/controller.go (hunk 10 above)
Comment (csrwng) [reply hypershift-operator/controllers/machinedeployment/controller.go]: It is also possible that the machinedeployment is for something other than a hypershift cluster.

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+
+	if err := r.Client.Patch(ctx, machineDeployment, originalMachineDeploymentToPatch); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch machineDeployment: %v", err)
+	}
+
+	return result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, machineDeployment *clusterv1.MachineDeployment) (ctrl.Result, error) {
+	klog.V(3).Infof("%v: Reconciling MachineDeployment", machineDeployment.Name)
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/controller.go]: please use the controller-runtime logger instead of klog
you can fetch the logger from the context with ctrl.LoggerFrom(ctx)

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Ignore deleted MachineDeployments, this can happen when foregroundDeletion
+	// is enabled
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/controller.go:282]: You should ignore any machinedeployments that are outside of a control plane namespace. I would fetch the namespace of the machinedeployments and immediately return if it doesn't have the `hypershift.openshift.io/hosted-control-plane: "true"` label

File: hypershift-operator/controllers/nodepool/capi.go [hunk 15]
```diff
@@ -707,17 +707,27 @@ func setMachineDeploymentReplicas(nodePool *hyperv1.NodePool, machineDeployment
...
+		// Guard: Enforce min=1 for non-AWS platforms to prevent scale-from-zero issues.
+		// Even if API validation fails or pre-existing objects exist with min=0, this prevents
+		// NodePools from being permanently stuck at 0 replicas on platforms that don't support
+		// scale-from-zero metadata.
+		effectiveMin := ptr.Deref(nodePool.Spec.AutoScaling.Min, 1)
```
Comment (muraee) [inline hypershift-operator/controllers/nodepool/capi.go]: why default to 1?

File: Containerfile.control-plane [hunk 16]
```diff
@@ -28,3 +28,7 @@ LABEL io.openshift.hypershift.control-plane-pki-operator-signs-csrs=true
 LABEL io.openshift.hypershift.hosted-cluster-config-operator-reports-node-count=true
 LABEL io.openshift.hypershift.control-plane-operator-supports-kas-custom-kubeconfig=true
 LABEL io.openshift.hypershift.control-plane-operator.v2-isdefault=true
+# TODO: Uncomment when CPO natively adds capacity annotations to MachineDeployments for scale-from-zero.
+# When uncommented, the hypershift-operator MachineDeployment controller will automatically skip reconciliation
+# and let CPO handle annotations natively.
+# LABEL io.openshift.hypershift.control-plane-operator-supports-aws-autoscaling-from-zero=true
```
Comment (muraee) [inline Containerfile.control-plane]: why do we need this? support from scale-from-zero is a CAPA feature, right? so it's related to the OCP release version of the HostedCluster not CPO capabilities

File: hypershift-operator/controllers/nodepool/capi.go (hunk 15 above)
Comment (jhjaggars) [reply hypershift-operator/controllers/nodepool/capi.go]: The rationale (at one point) is that 1 is the more common minimum value, but the guard that follows fixes things anyhow.  Will switch to 0 as that feels more reasonable.

File: Containerfile.control-plane (hunk 16 above)
Comment (jhjaggars) [reply Containerfile.control-plane]: We don't need this at this point.  The strategy has changed to inspecting the awsmachinetemplates for status.capacity rather than a label on the CPO.

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+	awsMachineTemplate, err := utils.ResolveAWSMachineTemplate(ctx, r.Client, machineDeployment)
+	if err != nil {
+		log.V(4).Info("Skipping MachineDeployment", "namespace", machineDeployment.Namespace, "name", machineDeployment.Name, "reason", err.Error())
+		// Not an error - this MachineDeployment may not be for AWS platform
+		return ctrl.Result{}, nil
+	}
+
+	// Check if AWSMachineTemplate.Status.Capacity is already populated by CPO
+	if len(awsMachineTemplate.Status.Capacity) > 0 {
+		log.V(3).Info("AWSMachineTemplate already has Status.Capacity set by CPO, skipping annotation workaround",
```
Comment (muraee) [inline hypershift-operator/controllers/machinedeployment/controller.go]: s/CPO/CAPA

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+	InstanceTypesCache InstanceTypesCache
+
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+// SetupWithManager creates a new controller for a manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineDeployment{}).
```
Comment (muraee) [inline hypershift-operator/controllers/machinedeployment/controller.go]: should add a predicate to only watch MDs with hypershift nodePool annotation `hypershift.openshift.io/nodePool`
see https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/nodepool/capi.go#L380

File: Containerfile.control-plane (hunk 16 above)
Comment (muraee) [reply Containerfile.control-plane]: ok, let's remove those comments then if it's not needed.

File: hypershift-operator/controllers/machinedeployment/utils/providerspec.go [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+const (
+	// RegionAnnotation is the fallback annotation for AWS region
+	RegionAnnotation = "capa.infrastructure.cluster.x-k8s.io/region"
+)
+
+// ResolveAWSMachineTemplate fetches the AWSMachineTemplate referenced by the MachineDeployment
+func ResolveAWSMachineTemplate(ctx context.Context, c client.Client, machineDeployment *clusterv1.MachineDeployment) (*infrav1.AWSMachineTemplate, error) {
+	// Extract infrastructureRef
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/utils/providerspec.go:37]: We should perform both of these checks (infrastructureRef.name and infraRef.kind) before attempting to fetch the namespace of the machinedeployments in the main controller body.

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+
+func (r *Reconciler) reconcile(ctx context.Context, machineDeployment *clusterv1.MachineDeployment) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(3).Info("Reconciling MachineDeployment", "machineDeployment", machineDeployment.Name)
+
+	// Resolve AWSMachineTemplate first
+	awsMachineTemplate, err := utils.ResolveAWSMachineTemplate(ctx, r.Client, machineDeployment)
+	if err != nil {
+		log.V(4).Info("Skipping MachineDeployment", "namespace", machineDeployment.Namespace, "name", machineDeployment.Name, "reason", err.Error())
+		// Not an error - this MachineDeployment may not be for AWS platform
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/controller.go]: If you move the ref checks (infra ref type and name) to the main controller loop, then an error resulting from fetching the AWSMachineTemplate should be retryable and should result in a reconcile error.

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+	originalMachineDeploymentToPatch := client.MergeFrom(machineDeployment.DeepCopy())
+
+	result, err := r.reconcile(ctx, machineDeployment)
+	if err != nil {
+		log.Error(err, "Failed to reconcile MachineDeployment")
+		r.recorder.Eventf(machineDeployment, corev1.EventTypeWarning, "ReconcileError", "%v", err)
+		// we don't return here so we want to attempt to patch the machine regardless of an error.
+	}
+
+	if err := r.Client.Patch(ctx, machineDeployment, originalMachineDeploymentToPatch); err != nil {
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/controller.go]: This patch shouldn't happen unconditionally. It should only happen if there is a difference from the orignal machinedeployments.

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+	awsClient, err := awsclient.NewValidatedClient(ctx, region, r.CredentialsFile, r.RegionCache)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error creating aws client: %w", err)
+	}
+
+	// Get instance type information
+	instanceTypeInfo, err := r.InstanceTypesCache.GetInstanceType(ctx, awsClient, region, instanceType)
+	if err != nil {
+		log.Error(err, "Unable to set scale from zero annotations: unknown instance type", "instanceType", instanceType)
+		log.Error(nil, "Autoscaling from zero will not work. To fix this, manually populate machine annotations for your instance type", "annotations", []string{cpuKey, memoryKey, gpuKey})
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/controller.go]: It seems this information might be more useful/visible in the event, rather than in the log.

File: hypershift-operator/controllers/machinedeployment/ec2_instance_types.go [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+func NewInstanceTypesCache() InstanceTypesCache {
+	cache := &instanceTypesCache{}
+	cache.cache = map[string]instanceTypesRegion{}
+	cache.rwmutex = sync.RWMutex{}
+	return cache
+}
+
+// GetInstanceType retrieves InstanceType from cache by name. If the cache is stale or nil it is refreshed first from the EC2 API.
+// The fetched instance types are specific to the region of the awsClient. Using region name as cacheID is recommended.
+func (i *instanceTypesCache) GetInstanceType(ctx context.Context, awsClient awsclient.Client, cacheID string, instanceType string) (InstanceType, error) {
```
Comment (csrwng) [inline hypershift-operator/controllers/machinedeployment/ec2_instance_types.go]: I think the code would be clearer if you just use region throughout instead of cacheID

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+	awsClient, err := awsclient.NewValidatedClient(ctx, region, r.CredentialsFile, r.RegionCache)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error creating aws client: %w", err)
+	}
+
+	// Get instance type information
+	instanceTypeInfo, err := r.InstanceTypesCache.GetInstanceType(ctx, awsClient, region, instanceType)
+	if err != nil {
+		r.recorder.Eventf(machineDeployment, corev1.EventTypeWarning, "UnknownInstanceType",
+			"Unable to set scale from zero annotations: unknown instance type %s. Autoscaling from zero will not work. To fix this, manually populate machine annotations for your instance type: %v",
```
Comment (gaol) [inline hypershift-operator/controllers/machinedeployment/controller.go]: > To fix this, manually populate machine annotations for your instance type

does this mean manual update on the MachineDeployment ? if so, I think it will lead to  reconciling again which may succeed ?

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+
+	// Resolve AWSMachineTemplate first
+	awsMachineTemplate, err := utils.ResolveAWSMachineTemplate(ctx, r.Client, machineDeployment)
+	if err != nil {
+		// Retryable error - could be transient API failure
+		return ctrl.Result{}, err
+	}
+
+	// Check if AWSMachineTemplate.Status.Capacity is already populated by CAPA
+	if len(awsMachineTemplate.Status.Capacity) > 0 {
```
Comment (gaol) [inline hypershift-operator/controllers/machinedeployment/controller.go:314]: maybe it is a silly question, but is it possible that the `awsMachineTemplate.Status.Capacity` gets updated after this reconciling ? 
how to test this case from QE side ?

Will setting an environment an accepted option instead of reading from runtime status to decide whether annotation workaround should be used ?

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 26]
```diff
@@ -0,0 +1,225 @@
...
+	InstanceTypesCache InstanceTypesCache
+
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+// SetupWithManager creates a new controller for a manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineDeployment{}, builder.WithPredicates(predicate.NewPredicateFuncs(r.hasNodePoolAnnotation))).
```
Comment (gaol) [inline hypershift-operator/controllers/machinedeployment/controller.go:61]: It is possible that a HC has multiple nodepools, some nodepools don't have auto-scaling enabled, shall we filter only on the machinedeployment where the corresponding nodepool enabled auto-scaling ?

File: hypershift-operator/main.go [hunk 27]
```diff
@@ -400,6 +420,20 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
 		return fmt.Errorf("unable to create controller: %w", err)
 	}
 
+	// Setup MachineDeployment controller for scale-from-zero
+	if opts.EnableScaleFromZero {
```
Comment (gaol) [inline hypershift-operator/main.go]: If this is only enabled on AWS now, shall we only set up the MachineDeployment controller if the platform is AWS ? At least there won't be miss leading logs saying the feature is enabled.

Comment (jhjaggars) [reply hypershift-operator/main.go]: Users can control if the controller starts by passing a valid credential or not.

File: hypershift-operator/controllers/machinedeployment/controller.go (hunk 26 above)
Comment (jhjaggars) [reply hypershift-operator/controllers/machinedeployment/controller.go:61]: It is possible, and likely.  It is harmless to annotate deployments with this information however.

File: hypershift-operator/controllers/machinedeployment/controller.go (hunk 25 above)
Comment (muraee) [reply hypershift-operator/controllers/machinedeployment/controller.go:314]: > maybe it is a silly question, but is it possible that the awsMachineTemplate.Status.Capacity gets updated after this reconciling ?

that could happen, I am thinking of an upgrade scenario from an older release to a release where CAPA supports scale-from-zero, where it will start setting this field.
I think we should remove the annotations if they exist here instead of just returning.

File: hypershift-operator/controllers/machinedeployment/controller.go (hunk 24 above)
Comment (jhjaggars) [reply hypershift-operator/controllers/machinedeployment/controller.go]: If the instance type is unknown, then I would expect reconciliation to fail again unless AWS added that instance type in the meantime.  In the case that it exists the second time then the controller will set the proper values, else it'll emit this event again.  I think the advice to manually set the annotations is probably bad advice since the instance type is probably incorrect.  I will update the event to remove that advice.

File: hypershift-operator/controllers/machinedeployment/controller.go (hunk 25 above)
Comment (jhjaggars) [reply hypershift-operator/controllers/machinedeployment/controller.go:314]: This is a good call out.  If the annotations are set and status.capacity is set, then the annotations will take precedence in the cluster-autoscaler.  This is backward from what we want.  I'm changing the controller to remove the annotations if status.capacity is set.

File: hypershift-operator/controllers/machinedeployment/controller.go (hunk 24 above)
Comment (jhjaggars) [reply hypershift-operator/controllers/machinedeployment/controller.go]: Further, I am making sure we retry on transient failures and just emit the event if the instance type is unknown.

File: test/e2e/autoscaling_test.go [hunk 28]
```diff
@@ -384,3 +385,254 @@ func newWorkLoad(njobs int32, memoryRequest resource.Quantity, nodeSelector, ima
 	}
 	return job
 }
+
+func TestNodePoolAutoscalingScaleFromZero(t *testing.T) {
+	if globalOpts.Platform != hyperv1.AWSPlatform {
+		t.Skip("test only supported on platform AWS")
+	}
+	if !globalOpts.HOInstallationOptions.EnableScaleFromZero {
```
Comment (csrwng) [inline test/e2e/autoscaling_test.go]: This may not be enough. The HOInstallationOptions are for tests that install the HO. The NodePool test uses the HO that is already installed (and configured in a previous workflow step).

You may need to fetch the HO deployment and inspect that.

File: hypershift-operator/main.go [hunk 29]
```diff
@@ -124,6 +126,8 @@ type StartOptions struct {
 	EnableUWMTelemetryRemoteWrite          bool
 	EnableValidatingWebhook                bool
 	EnableDedicatedRequestServingIsolation bool
+	EnableScaleFromZero                    bool
+	AWSCreds                               string // Path to mounted credentials file (set by deployment)
```
Comment (csrwng) [inline hypershift-operator/main.go]: These creds are specifically for the new controller, so I would name something specific like:
`scale-from-zero-aws-creds` or something similar.
(same for the flag name)

File: api/hypershift/v1beta1/nodepool_types.go [hunk 30]
```diff
@@ -446,11 +447,13 @@ type NodePoolManagement struct {
 // NodePoolAutoScaling specifies auto-scaling behavior for a NodePool.
 // +kubebuilder:validation:XValidation:rule="self.max >= self.min", message="max must be equal or greater than min"
 type NodePoolAutoScaling struct {
-	// min is the minimum number of nodes to maintain in the pool. Must be >= 1 and <= .Max.
+	// min is the minimum number of nodes to maintain in the pool.
+	// Can be set to 0 for scale-from-zero.
```
Comment (csrwng) [inline api/hypershift/v1beta1/nodepool_types.go]: `Can be set to 0 for scale-from-zero on AWS platform`

File: control-plane-operator/controllers/hostedcontrolplane/v2/assets/cluster-autoscaler/role.yaml (hunk 1 above)
Comment (sdminonne) [reply control-plane-operator/controllers/hostedcontrolplane/v2/assets/cluster-autoscaler/role.yaml:26]: Sure. From https://issues.redhat.com/browse/CNTRLPLANE-1622

> RBAC Updates
> 
>     Add watch verb to cluster-autoscaler RBAC for infrastructure.cluster.x-k8s.io API group
>     Allows informer to watch rather than poll for better performance

File: control-plane-operator/controllers/hostedcontrolplane/testdata/cluster-autoscaler/GCP/zz_fixture_TestControlPlaneComponents_cluster_autoscaler_role.yaml [hunk 31]
```diff
@@ -31,6 +31,7 @@ rules:
   verbs:
   - get
   - list
+  - watch
```
Comment (sdminonne) [inline control-plane-operator/controllers/hostedcontrolplane/testdata/cluster-autoscaler/GCP/zz_fixture_TestControlPlaneComponents_cluster_autoscaler_role.yaml:34]: Even if GCP won't support (yet) scale from zero.  Tough but fair

File: control-plane-operator/controllers/hostedcontrolplane/testdata/cluster-autoscaler/IBMCloud/zz_fixture_TestControlPlaneComponents_cluster_autoscaler_role.yaml [hunk 32]
```diff
@@ -31,6 +31,7 @@ rules:
   verbs:
   - get
   - list
+  - watch
```
Comment (sdminonne) [inline control-plane-operator/controllers/hostedcontrolplane/testdata/cluster-autoscaler/IBMCloud/zz_fixture_TestControlPlaneComponents_cluster_autoscaler_role.yaml:34]: Even if IBM won't support (yet) scale from zero. Tough but fair

File: hypershift-operator/controllers/machinedeployment/controller.go [hunk 33]
```diff
@@ -0,0 +1,305 @@
...
+		}
+
+		if changed {
+			log.Info("Removing workaround annotations since CAPA provides Status.Capacity",
+				"machineDeployment", machineDeployment.Name)
+			// Continue to patch the MachineDeployment with removed annotations
+		} else {
+			// No annotations to clean up, skip
+			return ctrl.Result{}, nil
+		}
```
Comment (muraee) [inline hypershift-operator/controllers/machinedeployment/controller.go:331]: "// Continue to patch the MachineDeployment with removed annotations"
this comment is confusing, please remove.

the `else`? is also unnecessary, we are returning anyway right after.

**Comment by jhjaggars** [comment]
/verified by @jhjaggars 

https://gist.github.com/jhjaggars/2ba3da57f5ad6929543106b78b4d16fd

**Comment by jhjaggars** [comment]
/verified by @jhjaggars

https://gist.github.com/jhjaggars/d4b286d1176c8c92fe0c5df2356b0da6

@gaol is also working on additional verification.
